src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs

```diff
@@ -1368,7 +1377,78 @@ public ValueTask<FlushResult> WritePipeAsync(ReadOnlyMemory<byte> data, Cancella
             else
             {
                 HandleNonBodyResponseWrite();
-                return !firstWrite ? default : Output.FlushAsync(cancellationToken);
+                return default;
+            }
+        }
+
+        private ValueTask<FlushResult> FirstWriteAsync(ReadOnlyMemory<byte> data, CancellationToken cancellationToken)
+        {
+            Debug.Assert(!HasResponseStarted);
+
+            var startingTask = FireOnStarting();
+
+            if (!ReferenceEquals(startingTask, Task.CompletedTask))
+            {
+                return FirstWriteAsyncAwaited(startingTask, data, cancellationToken);
+            }
+
+            var responseHeaders = InitializeResponseFirstWrite(data.Length);
```

Comment:
What would be the perf impact of moving every from this line down to its own method and returning its result e.g.
```csharp
return ThisBit();
```
Then the `FirstWriteAsyncAwaited` could become
```csharp
await initializeTask;
await ThisBit();
```
Which would cut down on code duplication

Reply:
Probably very little. I'll go ahead and do that.